src/font_file_lister_gdi.cpp

```diff
@@ -16,168 +16,130 @@
-
-font_index index_fonts(FontCollectorStatusCallback &cb) {
-	font_index hash_to_path;
-	auto fonts = get_installed_fonts();
-	std::unique_ptr<char[]> buffer(new char[1024]);
-	for (auto const& path : fonts) {
-		try {
-			auto stream = agi::io::Open(path, true);
-			stream->read(&buffer[0], 1024);
-			auto hash = murmur3(&buffer[0], stream->tellg());
-			hash_to_path.emplace(hash, path);
-		}
-		catch (agi::Exception const& e) {
-			cb(to_wx(e.GetMessage() + "\n"), 3);
-		}
+GdiFontFileLister::GdiFontFileLister(FontCollectorStatusCallback &)
+: dwrite_factory_sh(nullptr, [](IDWriteFactory* p) { p->Release(); })
+, font_collection_sh(nullptr, [](IDWriteFontCollection* p) { p->Release(); })
+, dc_sh(nullptr, [](HDC dc) { DeleteDC(dc); })
+, gdi_interop_sh(nullptr, [](IDWriteGdiInterop* p) { p->Release(); })
+{
+	IDWriteFactory* dwrite_factory;
+	if (FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), reinterpret_cast<IUnknown**>(&dwrite_factory))))
+	{
+		throw agi::EnvironmentError("Failed to initialize the DirectWrite Factory");
 	}
-	return hash_to_path;
-}
-
-void get_font_data(std::string& buffer, HDC dc) {
-	buffer.clear();
+	dwrite_factory_sh = dwrite_factory;
 
-	// For ttc files we have to ask for the "ttcf" table to get the complete file
-	DWORD ttcf = 0x66637474;
-	auto size = GetFontData(dc, ttcf, 0, nullptr, 0);
-	if (size == GDI_ERROR) {
-		ttcf = 0;
-		size = GetFontData(dc, 0, 0, nullptr, 0);
+	IDWriteFontCollection* font_collection;
+	if (FAILED(dwrite_factory_sh->GetSystemFontCollection(&font_collection, true)))
+	{
+		throw agi::EnvironmentError("Failed to initialize the system font collection");
 	}
-	if (size == GDI_ERROR || size == 0)
-		return;
+	font_collection_sh = font_collection;
 
-	buffer.resize(size);
-	GetFontData(dc, ttcf, 0, &buffer[0], size);
-}
-}
+	HDC dc = CreateCompatibleDC(nullptr);
+	if (dc == nullptr) {
+		throw agi::EnvironmentError("Failed to initialize the HDC");
+	}
+	dc_sh = dc;
 
-GdiFontFileLister::GdiFontFileLister(FontCollectorStatusCallback &cb)
-: dc(CreateCompatibleDC(nullptr), [](HDC dc) { DeleteDC(dc); })
-{
-	cb(_("Updating font cache\n"), 0);
-	index = index_fonts(cb);
+	IDWriteGdiInterop* gdi_interop;
+	if (FAILED(dwrite_factory_sh->GetGdiInterop(&gdi_interop)))
+	{
+		throw agi::EnvironmentError("Failed to initialize the Gdi Interop");
+	}
+	gdi_interop_sh = gdi_interop;
 }
 
 CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, int bold, bool italic, std::vector<int> const& characters) {
 	CollectionResult ret;
 
+	int weight = bold == 0 ? 400 :
+				 bold == 1 ? 700 :
+				 bold;
+
+	// From VSFilter
+	//   - https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/RTS.cpp#l45
+	//   - https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/STS.cpp#l2992
 	LOGFONTW lf{};
-	lf.lfCharSet = DEFAULT_CHARSET;
+	lf.lfCharSet = DEFAULT_CHARSET; // we should use the one specified in the ass file
```

Comment:
I wanted to talk about that. In some very specific situation, if we don't use the charset requested by the user, the selected font can be different then the font displayed by VSFilter.
But, this can hardly be emulated on macOS and Linux. So, do we really care to use the charset specified by the user? Even libass doesn't properly emulate how the charset works: https://github.com/libass/libass/issues/662

Reply:
Yes, it's probably better to treat encodings different from 1 as "unsupported". What I really meant here, though, is that currently this comment is a bit confusing. You can remove it or reword it to something like "note that this currently ignores the font encoding specified in the ass file".

Reply:
Addressed in commit 17a3f487f2479070be45fb454c44dc3dd8d724cc